### PR TITLE
update logic to disallow post cluster info contains spiffe_id

### PR DESCRIPTION
**Pull request by lixmgl**

### CODE REVIEW

File: deploy-board/deploy_board/webapp/cluster_view.py [hunk 1]
```diff
@@ -224,17 +224,9 @@ def post(self, request, name, stage):
...
-            log.info("getting current Cluster Configuration is {}", current_cluster)
-            if 'spiffe_id' not in current_cluster and 'spiffe_id' in cluster_info:
-                log.error("Teletraan does not support user to pass in customized spiffe_id %s if exising cluster does not have" % cluster_info['spiffe_id'])
-                raise TeletraanException("Teletraan does not support user to pass in customized spiffe_id")
-            if 'spiffe_id' in current_cluster and 'spiffe_id' in cluster_info:
-                if current_cluster['spiffe_id'] != cluster_info['spiffe_id']:
-                    log.error("Teletraan does not support user to change spiffe_id %s , current spiffe_id %s"
-                        % (cluster_info['spiffe_id'], current_cluster['spiffe_id']))
-                    raise TeletraanException("Teletraan does not support user to change spiffe_id")
+            if 'spiffe_id' in cluster_info:
```
Comment (yongwen) [inline deploy-board/deploy_board/webapp/cluster_view.py]: did you test the case that a normal update cluster configuration?  i thought in the normal case, the unchanged spiffe_id is also passed into the cluster_info and is required.

Comment (lixmgl) [reply deploy-board/deploy_board/webapp/cluster_view.py]: good catch, thanks. Please review updated change.

File: deploy-board/deploy_board/webapp/group_view.py [hunk 2]
```diff
@@ -30,6 +30,7 @@
 from helpers import autoscaling_metrics_helper
 from diff_match_patch import diff_match_patch
 from deploy_board import settings
+from helpers.exceptions import NotAuthorizedException, TeletraanException
```
Comment (euccas) [inline deploy-board/deploy_board/webapp/group_view.py]: Is "NotAuthorizedException" used somewhere?

Other changes look good to me.

File: deploy-board/deploy_board/webapp/group_view.py [hunk 3]
```diff
@@ -179,6 +180,11 @@ def update_launch_config(request, group_name):
             launchRequest["assignPublicIp"] = True
         else:
             launchRequest["assignPublicIp"] = False
+
+        if params["userData"].find("spiffe_id"):
```
Comment (yongwen) [inline deploy-board/deploy_board/webapp/group_view.py]: what if there was a spiffe_id in the userData and user did not change it but change other part of the userdata, we should still support this case.

File: deploy-board/deploy_board/webapp/cluster_view.py [hunk 4]
```diff
@@ -76,9 +76,10 @@ def post(self, request, name, stage):
             cluster_info = json.loads(request.body)
 
             log.info("Create Capacity in the provider")
-            if 'spiffe_id' in cluster_info:
-                log.error("Teletraan does not support user to change spiffe_id %s"  % cluster_info['spiffe_id'])
-                raise TeletraanException("Teletraan does not support user to create spiffe_id")
+            if 'configs' in cluster_info:
```
Comment (yongwen) [inline deploy-board/deploy_board/webapp/cluster_view.py]: looks like we change the behavior here, was it a bug before that spiffe_id supposed to be in cluster_info['config'] but not in cluster_info?

File: deploy-board/deploy_board/webapp/cluster_view.py [hunk 5]
```diff
@@ -224,17 +226,16 @@ def post(self, request, name, stage):
...
-                if current_cluster['spiffe_id'] != cluster_info['spiffe_id']:
-                    log.error("Teletraan does not support user to change spiffe_id %s , current spiffe_id %s"
-                        % (cluster_info['spiffe_id'], current_cluster['spiffe_id']))
-                    raise TeletraanException("Teletraan does not support user to change spiffe_id")
+            if 'configs' in current_cluster and 'configs' in cluster_info:
+                if 'spiffe_id' in current_cluster['configs'] and 'spiffe_id' in cluster_info['configs']:
+                    if current_cluster['configs']['spiffe_id'] != cluster_info['configs']['spiffe_id']:
+                       log.error("Teletraan does not support user to change spiffe_id %s" % cluster_info['spiffe_id'])
+                       raise TeletraanException("Teletraan does not support user to create spiffe_id")
+
```
Comment (yongwen) [inline deploy-board/deploy_board/webapp/cluster_view.py]: what if current_cluster has spiffe_id but the new cluster_info remove it, this logic will allow it through.

File: deploy-board/deploy_board/webapp/cluster_view.py [hunk 6]
```diff
@@ -224,17 +226,16 @@ def post(self, request, name, stage):
...
-            if 'spiffe_id' in current_cluster and 'spiffe_id' in cluster_info:
-                if current_cluster['spiffe_id'] != cluster_info['spiffe_id']:
-                    log.error("Teletraan does not support user to change spiffe_id %s , current spiffe_id %s"
-                        % (cluster_info['spiffe_id'], current_cluster['spiffe_id']))
-                    raise TeletraanException("Teletraan does not support user to change spiffe_id")
+            if 'configs' in current_cluster and 'configs' in cluster_info:
+                if 'spiffe_id' in current_cluster['configs'] and 'spiffe_id' in cluster_info['configs']:
+                    if current_cluster['configs']['spiffe_id'] != cluster_info['configs']['spiffe_id']:
+                       log.error("Teletraan does not support user to change spiffe_id %s" % cluster_info['spiffe_id'])
+                       raise TeletraanException("Teletraan does not support user to create spiffe_id")
```
Comment (yongwen) [inline deploy-board/deploy_board/webapp/cluster_view.py]: exception msg is different than the log message above.

File: deploy-board/deploy_board/webapp/group_view.py (hunk 2 above)
Comment (lixmgl) [reply deploy-board/deploy_board/webapp/group_view.py]: Yes, it is used by other existing functions. Thanks for reminding.

File: deploy-board/deploy_board/webapp/cluster_view.py (hunk 4 above)
Comment (lixmgl) [reply deploy-board/deploy_board/webapp/cluster_view.py]: Yes, it was a bug.

File: deploy-board/deploy_board/webapp/cluster_view.py (hunk 6 above)
Comment (lixmgl) [reply deploy-board/deploy_board/webapp/cluster_view.py]: updated, thanks for reminding.

File: deploy-board/deploy_board/webapp/cluster_view.py (hunk 5 above)
Comment (lixmgl) [reply deploy-board/deploy_board/webapp/cluster_view.py]: Good catch. Added this case.

File: deploy-board/deploy_board/webapp/group_view.py (hunk 3 above)
Comment (lixmgl) [reply deploy-board/deploy_board/webapp/group_view.py]: params is coming from request. I think this protection is good enough. 
The main purpose here is preventing user to update spiffe_id.
If data coming from user which is params does not contain spiffe_id, we are good.

File: deploy-board/deploy_board/webapp/group_view.py [hunk 7]
```diff
@@ -179,6 +180,30 @@ def update_launch_config(request, group_name):
...
+                if ":" in item:
+                    pair = item.split(":")
+                    if pair[0] == "spiffe_id" and userPassInSpiffeID is None:
+                        log.error("Teletraan does not support user to remove spiffe_id in userData field")
+                        raise TeletraanException("Teletraan does not support user to remove spiffe_id")
+
+                    if pair[0] == "spiffe_id" and pair[1] != userPassInSpiffeID:
+                        log.error("Teletraan does not support user to update spiffe_id in userData field")
+                        raise TeletraanException("Teletraan does not support user to update spiffe_id")
+
```
Comment (yongwen) [inline deploy-board/deploy_board/webapp/group_view.py]: looks like there is still a case that pair[0] != "spiffe_id" and userPassInSpiffeID

Comment (lixmgl) [reply deploy-board/deploy_board/webapp/group_view.py]: Nice catch :) Updated, thanks for pointing out.

File: deploy-board/deploy_board/webapp/group_view.py (hunk 2 above)
Comment (yongwen) [reply deploy-board/deploy_board/webapp/group_view.py]: looks like NotAuthorizedException is not used in this file.

**Pull request merged**